client/src/main/java/org/asynchttpclient/netty/channel/DefaultChannelPool.java

```diff
@@ -148,25 +153,15 @@ private boolean isIdleTimeoutExpired(IdleChannel idleChannel, long now) {
             return idleTimeoutChannels != null ? idleTimeoutChannels : Collections.<IdleChannel> emptyList();
         }
 
-        private boolean isChannelCloseable(Channel channel) {
-            Object attribute = Channels.getAttribute(channel);
-            if (attribute instanceof NettyResponseFuture) {
-                NettyResponseFuture<?> future = (NettyResponseFuture<?>) attribute;
-                if (!future.isDone()) {
-                    LOGGER.error("Future not in appropriate state {}, not closing", future);
-                    return false;
-                }
-            }
-            return true;
-        }
-
         private final List<IdleChannel> closeChannels(List<IdleChannel> candidates) {
 
             // lazy create, only if we have a non-closeable channel
             List<IdleChannel> closedChannels = null;
             for (int i = 0; i < candidates.size(); i++) {
+                // We call takeOwnership here to avoid closing a channel that has just been taken out
+                // of the pool, otherwise we risk closing an active connection.
                 IdleChannel idleChannel = candidates.get(i);
-                if (isChannelCloseable(idleChannel.channel)) {
+                if (idleChannel.takeOwnership()) {
```

Comment:
Well my very much uninformed opinion was that this check existed to prevent a channel with an incomplete future (e.g. one that is in use) from being closed, but this failed because that future is only registered when the request is sent. Normally if I haven't screwed up my change will prevent that from happening, so yes it's probably safe, unless there was some hidden meaning there that I didn't grasp...